File: packages/agent-harness/py/src/steerable_agent_harness/policy.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Literal


ToolMode = Literal["read", "safe_write", "destructive", "other"]


@dataclass(slots=True)
class PolicyDecision:
    allowed: bool
    tool_mode: ToolMode
    reason: str


#: Side-effect-free network reads the prefix rules cannot reach. Exact
#: names, not a ``web_`` prefix: a future ``web_deploy``-style tool must not
#: inherit the read posture. The approval algebra reads this table
#: (``ApprovalRequest.mode``), so the classification decides both the
#: prompt's risk label and the headless AutoApprover's default verdict.
_READ_EXACT = frozenset({"web_search", "web_fetch"})


def decide_tool_mode(tool_name: str) -> ToolMode:
    normalized = tool_name.lower()
    if normalized in _READ_EXACT:
        return "read"
    if normalized.startswith(("get_", "list_", "read_")):
        return "read"
    if normalized.startswith(("create_", "update_", "set_", "write_", "apply_")):
        return "safe_write"
    if normalized.startswith(("delete_", "drop_", "remove_", "destroy_")):
        return "destructive"
    return "other"

File: packages/agent-harness/ts/src/policy.ts
export type ToolMode = "read" | "safe_write" | "destructive" | "other";

export interface PolicyDecision {
  allowed: boolean;
  toolMode: ToolMode;
  reason: string;
}

// Side-effect-free network reads the prefix rules cannot reach. Exact
// names, not a "web_" prefix: a future web_deploy-style tool must not
// inherit the read posture. Kept in lockstep with the Python classifier
// (conformance case: tests/conformance/cases/policy/decide_tool_mode.yaml).
const READ_EXACT = new Set(["web_search", "web_fetch"]);

export function decideToolMode(toolName: string): ToolMode {
  const normalized = toolName.toLowerCase();
  if (READ_EXACT.has(normalized)) {
    return "read";
  }
  if (normalized.startsWith("get_") || normalized.startsWith("list_") || normalized.startsWith("read_")) {
    return "read";
  }
  if (
    normalized.startsWith("create_") ||
    normalized.startsWith("update_") ||
    normalized.startsWith("set_") ||
    normalized.startsWith("write_") ||
    normalized.startsWith("apply_")
  ) {
    return "safe_write";
  }
  if (
    normalized.startsWith("delete_") ||
    normalized.startsWith("drop_") ||
    normalized.startsWith("remove_") ||
    normalized.startsWith("destroy_")
  ) {
    return "destructive";
  }
  return "other";
}
